src/index/run.go

```diff
@@ -48,11 +48,17 @@ func (idx Index) Run(
 		checksumUpToDate = realChecksum == stg.Checksum
 	}
 
-	// Run if we have a command and no inputs.
-	doRun := hasCommand && (len(stg.Inputs) == 0)
+	var doRun bool
 
-	// Run if our checksum is stale.
-	doRun = doRun || !checksumUpToDate
+	if hasCommand && (len(stg.Inputs) == 0) {
+		// Run if we have a command and no inputs.
+		doRun = true
+		logger.Info.Printf("running stage %s because there is a command and no inputs\n", stagePath)
+	} else if !checksumUpToDate {
+		// Run if our checksum is stale.
+		doRun = true
+		logger.Info.Printf("running stage %s because the checksum is not up to date\n", stagePath)
+	}
```

Comment:
Looks like you missed a couple places where `doRun` is potentially set. ⬇️

Also I'd like to see some tests for this. I'd be fine with either traditional tests ([run_test.go](https://github.com/kevin-hanselman/dud/blob/8dcadb36ac3d7cf9b1b01ab41334d0639afb7e54/src/index/run_test.go#L71) 😄) or my ["integration" tests](https://github.com/kevin-hanselman/dud/tree/8dcadb36ac3d7cf9b1b01ab41334d0639afb7e54/integration/tests/basic_run), which are just shell scripts with some machinery built around them; peek around in `tests/integration/` to see some examples.